File: backend/src/main/java/ca/metricalsky/winston/entity/fetch/FetchActionEntity.java
package ca.metricalsky.winston.entity.fetch;

import jakarta.persistence.Basic;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.OffsetDateTime;

@Entity
@Table(name = "fetch_actions")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class FetchActionEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @Basic(optional = false)
    @Column(name = "fetch_operation_id")
    private Long fetchOperationId;

    @Basic(optional = false)
    @Enumerated(EnumType.STRING)
    @Column(name = "action_type")
    private Type actionType;

    @Basic(optional = false)
    @Column(name = "object_id")
    private String objectId;

    @Column(name = "published_after")
    private OffsetDateTime publishedAfter;

    @Column(name = "published_before")
    private OffsetDateTime publishedBefore;

    @Column(name = "page_token")
    private String pageToken;

    @Basic(optional = false)
    @Enumerated(EnumType.STRING)
    @Column(name = "status")
    private Status status;

    @Column(name = "item_count")
    private Integer itemCount;

    @Column(name = "error")
    private String error;

    @CreationTimestamp
    @Column(name = "created_at")
    private OffsetDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "last_updated_at")
    private OffsetDateTime lastUpdatedAt;

    public enum Type {
        CHANNELS,
        VIDEOS,
        COMMENTS,
        REPLIES,
    }

    public enum Status {
        READY,
        FETCHING,
        SUCCESSFUL,
        FAILED,
    }
}

File: database/src/main/kotlin/ca/metricalsky/winston/database/entity/fetch/FetchActionEntity.kt
package ca.metricalsky.winston.database.entity.fetch

import jakarta.persistence.Basic
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import java.time.OffsetDateTime


@Entity
@Table(name = "fetch_actions")
class FetchActionEntity(

    @Basic(optional = false)
    @Column(name = "fetch_operation_id")
    var fetchOperationId: Long? = null,

    @Basic(optional = false)
    @Enumerated(EnumType.STRING)
    @Column(name = "action_type")
    var actionType: Type,

    @Basic(optional = false)
    @Column(name = "object_id")
    var objectId: String,

    @Column(name = "published_after")
    var publishedAfter: OffsetDateTime? = null,

    @Column(name = "published_before")
    var publishedBefore: OffsetDateTime? = null,
) {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    var id: Long? = null

    @Column(name = "page_token")
    var pageToken: String? = null

    @Basic(optional = false)
    @Enumerated(EnumType.STRING)
    @Column(name = "status")
    var status = Status.READY

    @Column(name = "item_count")
    var itemCount: Int? = null

    @Column(name = "error")
    var error: String? = null

    @CreationTimestamp
    @Column(name = "created_at")
    var createdAt: OffsetDateTime? = null

    @UpdateTimestamp
    @Column(name = "last_updated_at")
    var lastUpdatedAt: OffsetDateTime? = null

    enum class Type {
        CHANNELS,
        VIDEOS,
        COMMENTS,
        REPLIES,
    }

    enum class Status {
        READY,
        FETCHING,
        SUCCESSFUL,
        FAILED,
    }
}
